numpy/lib/format.py

```diff
@@ -915,3 +919,130 @@ def _read_bytes(fp, size, error_template="ran out of data"):
         raise ValueError(msg % (error_template, size, len(data)))
     else:
         return data
+
+class AppendArray:
+    def __init__(self, filename):
+        self.filename = filename
+        self.fp = None
+        self.__is_init = False
+        if os.path.isfile(filename):
+            self.__init()
+
+    def __create_header_bytes(self):
+        header_map = {
+            'descr': dtype_to_descr(self.dtype),
+            'fortran_order': self.fortran_order,
+            'shape': tuple(self.shape)
+        }
+        io = BytesIO()
+        write_array_header_2_0(io, header_map)
+
+        # create array header with 64 byte space space for shape to grow
+        io.getbuffer()[8:12] = pack("<I", int(io.getbuffer().nbytes-12+64))
+        io.getbuffer()[-1] = 32
+        io.write(b" "*64)
+        io.getbuffer()[-1] = 10
+
+        return io.getbuffer()
+
+    def __init(self, arr = None):
+        self.fp = open(self.filename, mode="rb+" if arr is None else "wb")
+        fp = self.fp
+
+        if arr is None:
+            magic = read_magic(fp)
+
+            if magic != (2, 0):
+                raise NotImplementedError(
+                    "version (%d, %d) not implemented" % magic
+                )
+
+            header = read_array_header_2_0(fp)
+            shape, self.fortran_order, self.dtype = header
+            self.shape = list(shape)
+
+            if self.fortran_order == True:
+                raise NotImplementedError("fortran_order not implemented")
+
+            self.header_length = fp.tell()
+
+            header_length = self.header_length
+
+            new_header_bytes = self.__create_header_bytes()
+
+            if len(new_header_bytes) != header_length:
+                raise TypeError("no spare header space in target file %s" % (
+                    self.filename
+                ))
+
+            self.fp.seek(0, SEEK_END)
+
+        else:
+            self.shape, self.fortran_order, self.dtype = \
+                list(arr.shape), False, arr.dtype
+
+            fp.write(self.__create_header_bytes())
+
+            self.header_length = fp.tell()
+
+            arr.tofile(fp)
+
+        self.__is_init = True
+
+    def append(self, arr):
+        if not arr.flags.c_contiguous:
+            raise NotImplementedError("ndarray needs to be c_contiguous")
+
+        if not self.__is_init:
+            self.__init(arr)
+            return
+
+        if arr.dtype != self.dtype:
+            raise TypeError("incompatible ndarrays types %s and %s" % (
+                arr.dtype, self.dtype
+            ))
+
+        shape = self.shape
+
+        if len(arr.shape) != len(shape):
+            raise TypeError("incompatible ndarrays shape lengths %s and %s" % (
+                len(arr.shape), len(shape)
+            ))
+
+        if shape[1:] != list(arr.shape[1:]):
+            raise TypeError("ndarray shapes can only differ on zero axis")
+
+        self.shape[0] += arr.shape[0]
+
+        arr.tofile(self.fp)
```

Comment:
This doesn't quite feel safe to me; you can't guarantee in python that the `__del__` or `close` method is actually called, and you can risk leaving behind a silently truncated file without any record that it got truncated.

Should we have a "dirty" flag in the header, which gets cleared by `close()` such that the user can at least detect when they open a file that's broken in this way?

Reply:
This issue used to be around before as well, also in numpy.save: if for some reason the program crashes before the file is finished, it header and content will be in an inconsistent state. This was also an issue in all previous versions of AppendArray but may be more pronounced here.

Did I get you right that you supposed to add a dirty flag to the .npy file format? I was trying to add some functionality without introducing a new numpy file format version. However, I already had some ideas for a follow up pull request, which would allow -1 to be specified as shape[0], so that the size of the array could be infered from the file size. This would be consistent with e.g. numpy.reshape. Would that eliminate the need for a dirty flag, what do you think?

Reply:
So, here are the options as I see them:
* Rewrite the header after every append call. This probably slows things down a bit as you end up jumping back and forth in the file. This at least means that as long as `append` itself is not interrupted, the user knows their data is safe. Perhaps we should profile it.
* Write `-1` to the header while the `AppendArray` is open, then replace it with the final size when `.close()` runs normally. This is essentially the dirty flag I was suggesting, without needing any new header space. Old versions of numpy will not be able to open a file that is dirty in this way.
* Add `"dirty": True` to the header dictionary while the `AppendArray` is open, and then remove it again

Reply:
Option 2, the -1 as shape[0] of the array I have meant in a different way: the -1 would basically stay there forever, the header would not need to be touched anymore and would also never grow. Having the -1 there indicates that the size of the array is determined by the file size. So when the array is loaded and the actual ndarray constructed, the shape of the ndarray would not contain -1 but the actual size in the ndarray object, while the .npy file header stays the same. This would allow .npy to be the digital version of CSV (just kidding, I mean "binary version of CSV" of course), where data is also simply appended to the file's end. Perfectly suited for binary log files.
There would not be any indication whether a write has succeeded or failed, though. A partial, unsuccessful write could however be potentially identified if the file size (minus header size) is not divisible by the multiplied shape entries (times entry byte size). So potential errors need to be caught elsewhere in the file creation process.
All .npy files could theoretically have their shape[0] of -1, since it should be possible to derive the array size from the file size anyway.
What do you think about this interpretation of option 2?

I cannot speak for fortran_order though. Does anybody have a clue?

Reply:
I have implemented option 1, "Rewrite the header after every append call" and then option 2 as described in a later commit. What do you think? I can revert to the option 1 commit if option 2 goes too far, as it slightly changes the .npy format.

Also, I made some assumptions about how the fortran part might work. If someone had a fortran array to actually test `AppendArray`, that would be great.